compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/InitializersLowering.kt

```diff
@@ -7,84 +7,72 @@ package org.jetbrains.kotlin.backend.common.lower
-    context: CommonBackendContext
-) : InitializersLoweringBase(context) {
+open class InitializersLowering(context: CommonBackendContext) : InitializersLoweringBase(context) {
     override fun lower(irClass: IrClass) {
-        val nonStaticDeclarations = getDeclarationsWithNonStaticInitializers(irClass)
-        val instanceInitializerStatements = nonStaticDeclarations.mapNotNull { handleDeclaration(irClass, it) }
-        transformInstanceInitializerCallsInConstructors(irClass, instanceInitializerStatements)
-
-        val anonymousInitializers = nonStaticDeclarations.filterTo(hashSetOf()) { it is IrAnonymousInitializer }
-        irClass.declarations.removeAll(anonymousInitializers)
-    }
-
-    private fun getDeclarationsWithNonStaticInitializers(irClass: IrClass): List<IrDeclaration> =
-        irClass.declarations.filter {
+        val instanceInitializerStatements = extractInitializers(irClass) {
             (it is IrField && !it.isStatic) || (it is IrAnonymousInitializer && !it.isStatic)
         }
+        irClass.transformChildrenVoid(object : IrElementTransformerVoidWithContext() {
+            // Only transform constructors of current class.
+            override fun visitClassNew(declaration: IrClass) = declaration
 
-    private fun transformInstanceInitializerCallsInConstructors(irClass: IrClass, instanceInitializerStatements: List<IrStatement>) {
-        irClass.transformChildrenVoid(object : IrElementTransformerVoid() {
-            override fun visitInstanceInitializerCall(expression: IrInstanceInitializerCall): IrExpression {
-                val copiedBlock =
-                    IrBlockImpl(UNDEFINED_OFFSET, UNDEFINED_OFFSET, context.irBuiltIns.unitType, null, instanceInitializerStatements)
-                        .copy(irClass) as IrBlock
-                return IrBlockImpl(irClass.startOffset, irClass.endOffset, context.irBuiltIns.unitType, null, copiedBlock.statements)
-            }
+            override fun visitSimpleFunction(declaration: IrSimpleFunction) = declaration
+
+            override fun visitInstanceInitializerCall(expression: IrInstanceInitializerCall): IrExpression =
+                IrBlockImpl(irClass.startOffset, irClass.endOffset, context.irBuiltIns.unitType, null, instanceInitializerStatements)
+                    .deepCopyWithSymbols(currentScope!!.scope.getLocalDeclarationParent())
         })
     }
 }
 
-abstract class InitializersLoweringBase(val context: CommonBackendContext) : ClassLoweringPass {
-    protected fun handleDeclaration(irClass: IrClass, declaration: IrDeclaration): IrStatement? = when (declaration) {
-        is IrField -> handleField(irClass, declaration)
-        is IrAnonymousInitializer -> handleAnonymousInitializer(declaration)
-        else -> null
+abstract class InitializersLoweringBase(open val context: CommonBackendContext) : ClassLoweringPass {
+    protected fun extractInitializers(irClass: IrClass, filter: (IrDeclaration) -> Boolean): List<IrStatement> {
```

Comment:
I'd prefer to make it more readable, something like this, what do you think?
```
    protected fun extractInitializers(irClass: IrClass, filter: (IrDeclaration) -> Boolean): List<IrStatement> =
        irClass.declarations.filter(filter).mapNotNull {
            when (it) {
                is IrField -> handleField(irClass, it)
                is IrAnonymousInitializer -> handleAnonymousInitializer(it)
                else -> null
            }
        }.also { 
            irClass.declarations.removeAll { it is IrAnonymousInitializer }
        }
```

Reply:
Sure. My intention was to not have any list copies, but I guess that's my C++ background making itself visible. (Though note that your snippet is slightly incorrect - `removeAll` should also check `filter(it)`.)